cameleon/src/u3v/stream_handle.rs

```diff
@@ -232,45 +189,93 @@ impl StreamingLoop {
                 },
             };
 
-            let leader = match read_leader(&mut inner, &self.params, &mut leader_buf) {
-                Ok(leader) => leader,
-                Err(err) => {
-                    // Report and send error if the error is fatal.
-                    if matches!(err, StreamError::Io(..) | StreamError::Disconnected) {
-                        error!(?err);
-                        self.sender.try_send(Err(err)).ok();
-                    }
-                    payload_buf_opt = Some(payload_buf);
-                    continue;
+            let mut async_pool = AsyncPool::new(&inner);
+
+            if let Err(err) = read_leader(&mut async_pool, &self.params, &mut leader_buf) {
+                // Report and send error if the error is fatal.
+                if matches!(err, StreamError::Io(..) | StreamError::Disconnected) {
+                    error!(?err);
+                    self.sender.try_send(Err(err)).ok();
                 }
+                payload_buf_opt = Some(payload_buf);
```

Comment:
Yes there are probably some good opportunities for simplification here. As you said, let's keep that for another PR because I want to change the logic here as little as possible compared to what it was.